File: MaxCounters/ruby/MaxCounters.rb
class MaxCounters

  def solution(n, a)

    max = 0
    last = 0
    temp_array = Array.new(n, 0)

    a.each do |item|
      if (item > n)
        # get max
        max = last
      else
        # increase
        if (max > temp_array[item-1])
          # if max is higher
          temp_array[item-1] = max + 1    
        else
          temp_array[item-1] += 1
        end

        # get last if it's higher
        if (temp_array[item-1] > last)
          last = temp_array[item-1]
        end
      end
    end

    # fix lower than max to last max
    temp_array.each_with_index do |item, index|
      temp_array[index] = max if (item < max)
    end

    temp_array

  end

end

File: MaxCounters/swift/MaxCounters.swift
func solution(n: Int, a: [Int]) -> [Int] {

	var max = 0
	var last = 0
	var temp_array = Array(count: n, repeatedValue: 0)

	for item in a 
	{
		if item > n	{

			max = last

		} else {

			if max > temp_array[item-1] {
				temp_array[item-1] = max + 1
			} else {
				temp_array[item-1] += 1
			}

			if temp_array[item-1] > last {
				last = temp_array[item-1]
			}

		}

	}

	for (index, item) in enumerate(temp_array)
	{
		if item < max {
			temp_array[index] = max
		}
	}

	return temp_array

}

println(solution(5,[3, 4, 4, 6, 1, 4, 4]))